### ISSUE TITLE
DistAutogradTest.TestInitializedContextCleanup and DistAutogradTest.TestInitializedContextCleanupSendFunction fail

### ISSUE THREAD

**Issue by deepali-c** [pytorch/pytorch#38710]
## 🐛 Bug
Observed the following test to fail:
```
DistAutogradTest.TestInitializedContextCleanup
DistAutogradTest.TestInitializedContextCleanupSendFunction
```

with the following error:
`C++ exception with description "Already have an autograd context id for this thread.`

1. Build PyTorch v1.5.0 with `BUILD_TEST` and `INSTALL_TEST` enabled. 
2. Execute `build/bin/test_dist_autograd`

Observed output:
```
$ ./test_dist_autograd
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from DistAutogradTest
[ RUN      ] DistAutogradTest.TestSendFunctionInvalidInputs
[       OK ] DistAutogradTest.TestSendFunctionInvalidInputs (1 ms)
[ RUN      ] DistAutogradTest.TestInitializedContextCleanup
unknown file: Failure
C++ exception with description "Already have an autograd context id for this thread. (newContext at <mycode>/work/torch/csrc/distributed/autograd/context/container.cpp:98)
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xc0 (0x7fff97827c30 in <myenv>/lib/python3.6/site-packages/torch/lib/libc10.so)
frame #1: torch::distributed::autograd::DistAutogradContainer::newContext() + 0x97c (0x7fffb2efc53c in <myenv>/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so)
frame #2: torch::distributed::autograd::DistAutogradTest_TestInitializedContextCleanup_Test::TestBody() + 0x38 (0x10f2dff98 in ./test_dist_autograd)
frame #3: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x7c (0x10f32365c in ./test_dist_autograd)
frame #4: testing::Test::Run() + 0x134 (0x10f30d044 in ./test_dist_autograd)
frame #5: testing::TestInfo::Run() + 0x1f8 (0x10f30d2b8 in ./test_dist_autograd)
frame #6: testing::TestCase::Run() + 0x13c (0x10f30d60c in ./test_dist_autograd)
frame #7: testing::internal::UnitTestImpl::RunAllTests() + 0x5c0 (0x10f316e20 in ./test_dist_autograd)
frame #8: testing::UnitTest::Run() + 0xc4 (0x10f3170c4 in ./test_dist_autograd)
frame #9: main + 0x108 (0x10f2dd7e8 in ./test_dist_autograd)
frame #10: <unknown function> + 0x25280 (0x7fff973f5280 in /lib64/libc.so.6)
frame #11: __libc_start_main + 0xc4 (0x7fff973f5474 in /lib64/libc.so.6)
" thrown in the test body.
[  FAILED  ] DistAutogradTest.TestInitializedContextCleanup (1 ms)
[ RUN      ] DistAutogradTest.TestInitializedContextCleanupSendFunction
unknown file: Failure
C++ exception with description "Already have an autograd context id for this thread. (newContext at <mycode>/work/torch/csrc/distributed/autograd/context/container.cpp:98)
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xc0 (0x7fff97827c30 in <myenv>/lib/python3.6/site-packages/torch/lib/libc10.so)
frame #1: torch::distributed::autograd::DistAutogradContainer::newContext() + 0x97c (0x7fffb2efc53c in <myenv>/lib/python3.6/site-packages/torch/lib/libtorch_cpu.so)
frame #2: torch::distributed::autograd::DistAutogradTest_TestInitializedContextCleanupSendFunction_Test::TestBody() + 0x58 (0x10f2e07a8 in ./test_dist_autograd)
frame #3: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) + 0x7c (0x10f32365c in ./test_dist_autograd)
frame #4: testing::Test::Run() + 0x134 (0x10f30d044 in ./test_dist_autograd)
frame #5: testing::TestInfo::Run() + 0x1f8 (0x10f30d2b8 in ./test_dist_autograd)
frame #6: testing::TestCase::Run() + 0x13c (0x10f30d60c in ./test_dist_autograd)
frame #7: testing::internal::UnitTestImpl::RunAllTests() + 0x5c0 (0x10f316e20 in ./test_dist_autograd)
frame #8: testing::UnitTest::Run() + 0xc4 (0x10f3170c4 in ./test_dist_autograd)
frame #9: main + 0x108 (0x10f2dd7e8 in ./test_dist_autograd)
frame #10: <unknown function> + 0x25280 (0x7fff973f5280 in /lib64/libc.so.6)
frame #11: __libc_start_main + 0xc4 (0x7fff973f5474 in /lib64/libc.so.6)
" thrown in the test body.
[  FAILED  ] DistAutogradTest.TestInitializedContextCleanupSendFunction (0 ms)
[----------] 3 tests from DistAutogradTest (2 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (2 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] DistAutogradTest.TestInitializedContextCleanup
[  FAILED  ] DistAutogradTest.TestInitializedContextCleanupSendFunction
```

## Expected behavior
The tests are expected to pass. 

## Environment
```
$ python collect_env.py
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A

OS: Red Hat Enterprise Linux Server 7.7 (Maipo)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
CMake version: version 3.14.0

Python version: 3.6
Is CUDA available: N/A
CUDA runtime version: 10.2.89
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
```

## Additional context

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @xush6528 @jjlilley @osalpekar

**Comment by mrshenli** [comment]
Hey @deepali-c, thanks for trying out RPC. This issue might have been fixed by #38512, can you try if the error still exist after that PR?

BTW, does this test consistently fail in your env or flaky?

**Comment by deepali-c** [comment]
@mrshenli, yes this is a consistent failure that I see.  

I think it is just a test case issue. As in, in the DistAutogradTest class, DistAutogradContainer is created static in SetUpTestCase  to be shared by the three tests in the class.  So the first test goes fine, but in the beginning of the 2nd test execution newContext() fails because the DistAutogradContainer is not released at the end of the first test.  

I tried to build pytorch v1.5.0 with fix from #38512, but looks like some more changes went in between that are needed to get it to build successfully.

**Comment by pritamdamania87** [comment]
Closing out since https://github.com/pytorch/pytorch/pull/38711 has landed.
